File: 93. Restore IP Addresses/Solution.java
class Solution {
    List<String> res;
    int n, l;
    List<StringBuilder> strs;
    public List<String> restoreIpAddresses(String s) {
        res = new ArrayList<>();
        n = s.length();
        l = 0;
        strs = new ArrayList<>();
        strs.add(new StringBuilder());
        strs.add(new StringBuilder());
        strs.add(new StringBuilder());
        strs.add(new StringBuilder());
        backtrack(s, 0, 0);
        return res;
    }
    public void backtrack(String s, int curr, int pointer) {
        if (curr == n) {
            if (pointer < 3) return;
            StringBuilder resBuilder = new StringBuilder();
            for (StringBuilder str : strs) {
                if (str.length() == 0) return;
                if (str.charAt(0) == '0' && str.length() > 1) return;
                resBuilder.append(str.toString());
                resBuilder.append('.');
            }
            resBuilder.deleteCharAt(resBuilder.length() - 1);
            res.add(resBuilder.toString());
            return;
        }
        strs.get(pointer).append(s.charAt(curr));
        if (Integer.parseInt(strs.get(pointer).toString()) <= 255) {
            backtrack(s, curr + 1, pointer);
            if (pointer < 3) backtrack(s, curr + 1, pointer + 1);
        }
        strs.get(pointer).deleteCharAt(strs.get(pointer).length() - 1);
    }
}

File: 93. Restore IP Addresses/Solution.py
class Solution:
    def restoreIpAddresses(self, s: str) -> List[str]:
        self.res = []
        self.strs = ["", "", "", ""]
        self.n = len(s)
        self.backtrack(s, 0, 0)
        return self.res
        
    def backtrack(self, s, curr, pointer):
        if curr == self.n:
            if pointer < 3: return
            for s in self.strs:
                if len(s) == 0: return
                if s[0] == '0' and len(s) > 1: return
            self.res.append('.'.join(self.strs))
            return
        self.strs[pointer] += s[curr]
        if int(self.strs[pointer]) <= 255:
            self.backtrack(s, curr + 1, pointer)
            if pointer < 3: self.backtrack(s, curr + 1, pointer + 1)
        self.strs[pointer] = self.strs[pointer][:-1]